src/Libraries/DSCPython/CPythonEvaluator.cs

```diff
@@ -151,10 +149,78 @@ public static class CPythonEvaluator
         private const string NodeName = "__dynamonodename__";
         static PyScope globalScope;
         internal static readonly string globalScopeName = "global";
+        internal static DynamoLogger DynamoLogger {
+            get
+            { // Session is null when running unit tests.
+                if (ExecutionEvents.ActiveSession != null)
+                {
+                    return ExecutionEvents.ActiveSession.GetParameterValue(ParameterKeys.Logger) as DynamoLogger;
+                }
+                return null;
+            }
+        }
 
         static CPythonEvaluator()
         {
             InitializeEncoders();
+            Dynamo.Models.DynamoModel.RequestPythonReset += RequestPythonRestartHandler;
+           
+        }
+
+        internal static void RequestPythonRestartHandler(string pythonEngine)
+        {
+            //check if python engine request is for this engine, and engine is currently started
+            if (PythonEngine.IsInitialized && pythonEngine == "CPython3")
+            {
+                DynamoLogger?.Log("attempting reload of cpython3 modules", LogLevel.Console);
+                using (Py.GIL())
+                {
+                //the following is inspired by:
+                //https://github.com/ipython/ipython/blob/master/IPython/extensions/autoreload.py
+                    var global = PyScopeManager.Global.Get(CPythonEvaluator.globalScopeName);
+                    global?.Exec(@"import sys
+import importlib
+import importlib.util
+import os
+def getInfoFile(module):
+    if not hasattr(module, '__file__') or module.__file__ is None:
+        return None
+
+    if getattr(module, '__name__', None) in [None, '__mp_main__', '__main__']:
+        # we cannot reload(__main__) or reload(__mp_main__)
+        return None
+
+    filename = module.__file__
+    path, ext = os.path.splitext(filename)
+
+    if ext.lower() == '.py':
+        py_filename = filename
+    else:
+        try:
+            py_filename = importlib.util.source_from_cache(filename)
+        except ValueError:
+            return None
+    return py_filename
+
+
+for modname,mod in sys.modules.copy().items():
```

Comment:
any idea why we need to make the copy() call ?

Reply:
yes because iterating over sys.modules and reloading modules directly modifies the keys in this dictionary during iteration.

Making a copy avoids modifying the dictionary we're iterating over.